### PR TITLE
 Bug #120430 fix: File type acceptance issue, upload & download (uploaded file), new param for file size

### PR DESCRIPTION
**Pull request by vaibhavsTekdi**
File type acceptance issue, upload & download (uploaded file), new param for file size

### CODE REVIEW

File: administrator/models/field.php [hunk 1]
```diff
@@ -172,6 +172,7 @@ public function save_option($post)
 	{
 		$table = $this->getTable();
 		$data  = $post->get('jform', '', 'ARRAY');
+
```
Comment (ankush-maherwal) [inline administrator/models/field.php]: remove extra line

File: admin_language/en-GB/en-GB.com_tjfields.ini [hunk 2]
```diff
@@ -134,7 +134,7 @@ COM_TJFIELDS_FORM_LBL_EDITOR_HEIGHT="Height"
 COM_TJFIELDS_FORM_LBL_EDITOR_HEIGHT="Set height"
 COM_TJFIELDS_FORM_DESC_EDITOR_HEIGHT="Default Value"
 COM_TJFIELDS_FORM_LBL_FIELD_SIZE="Size"
-COM_TJFIELDS_FORM_DESC_FIELD_SIZE="Set size"
+COM_TJFIELDS_FORM_DESC_FIELD_SIZE="Set file size in MB (Numbers Only)"
```
Comment (ankush-maherwal) [inline admin_language/en-GB/en-GB.com_tjfields.ini]: Set maximum allowed file size (In MB) e.g 10, 20

File: admin_language/en-GB/en-GB.com_tjfields.ini [hunk 3]
```diff
@@ -385,3 +385,12 @@ COM_TJFIELDS_PERMISSION_EDIT_FIELD_VALUE_DESC="Allows users in the group to edit
 COM_TJFIELDS_PERMISSION_EDIT_OWN_FIELD_VALUE="Edit Own Field Value"
 COM_TJFIELDS_PERMISSION_EDIT_OWN_FIELD_VALUE_DESC="Allows users in the group to edit own data for the field"
 COM_TJFIELDS_FILE_DOWNLOAD="Download"
+
+;textareacounter field
+COM_TJFIELDS_TEXTAREACOUNTER="Textareacounter"
+COM_TJFIELDS_FORM_LBL_FIELD_COUNTERTEXT="Contertext"
+COM_TJFIELDS_FORM_DESC_FIELD_COUNTERTEXT="Format : Characters remaining {remaining} / {maxlength}"
+COM_TJFIELDS_FORM_LBL_FIELD_MAXLENGHT="Maxlength"
+COM_TJFIELDS_FORM_DESC_FIELD_MAXLENGHT="The maximum number of characters allowed in the text area"
```
Comment (ankush-maherwal) [inline admin_language/en-GB/en-GB.com_tjfields.ini]: Check the constants (Seems that file if not commited)

File: administrator/models/fields/file.php [hunk 4]
```diff
@@ -128,6 +128,7 @@ protected function getInput()
 	{
 		$layoutData = $this->getLayoutData();
 		$html = '';
+		$html = $this->getRenderer($this->layout)->render($layoutData);
```
Comment (ankush-maherwal) [inline administrator/models/fields/file.php]: check the layout override with bootstrap 3 template  (File and Textarea counter)

File: site/controller.php [hunk 5]
```diff
@@ -42,19 +51,24 @@ public function getMedia()
...
+
+		$explodeFilePath = explode('/', $decodedPath);
+		$fileName = end($explodeFilePath);
+
 		$db = JFactory::getDbo();
 		$query = $db->getQuery(true);
 		$query->select('*');
 		$query->from('#__tjfields_fields_value');
-		$query->quoteName('id') . ' = ' . $db->quote($file_id);
+		$query->where($db->quoteName('value') . " LIKE " . $db->Quote('%' . $fileName . '%'));
```
Comment (ankush-maherwal) [inline site/controller.php]: need an exact match, LIKE can cause uncertainties in the code

File: site/controller.php [hunk 6]
```diff
@@ -65,23 +79,22 @@ public function getMedia()
...
+				$app->redirect($this->returnURL);
 			}
 
 			return;
 		}
 		else
 		{
 			$app->enqueueMessage(JText::_('JERROR_ALERTNOAUTHOR'), 'error');
-			$app->redirect($uri);
+			$app->redirect($this->returnURL);
```
Comment (ankush-maherwal) [inline site/controller.php]: do we need redirection in ajax call?

File: site/helpers/tjfields.php [hunk 7]
```diff
@@ -205,20 +204,24 @@ public function saveFieldsValue($data)
 
 							if ($insert_obj_file->value)
 							{
-								if (!empty($if_edit_file_id))
+								if ($if_edit_file_id)
 								{
 									$insert_obj_file->id = $if_edit_file_id;
-									$db->updateObject('#__tjfields_fields_value', $insert_obj_file, 'id');
+									$result = $db->updateObject('#__tjfields_fields_value', $insert_obj_file, 'id');
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: $result never used

File: site/helpers/tjfields.php [hunk 8]
```diff
@@ -277,7 +280,7 @@ public function saveFieldsValue($data)
 
 							$db->setQuery($query);
 
-							$db->execute();
+							$result = $db->execute();
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: $result never used

File: site/helpers/tjfields.php [hunk 9]
```diff
@@ -401,19 +404,50 @@ public function uploadFile($singleFile, $insert_obj_file, $file_field_data)
...
 			$fileSize = $singleFile['size'];
 
-			if ($fileSize > 15728640)
+			if($acceptSize != "")
 			{
-				$app->enqueueMessage('This <strong>' . $title . '</strong> file size is greater than 15MB', 'warning');
+				$filesizeInBytes = $this->formatSizeUnits($acceptSize);
+				if($fileSize > $filesizeInBytes)
+				{
+					$app->enqueueMessage('File size is greater than '. $acceptSize . 'MB' , 'warning');
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: use language constant

File: site/helpers/tjfields.php [hunk 10]
```diff
@@ -401,19 +404,50 @@ public function uploadFile($singleFile, $insert_obj_file, $file_field_data)
...
+
+			$fieldId = (int) $file_field_data->id;
+			$fieldItems = $fieldModel->getItem($fieldId);
+
+			// code for file size validation
+			$acceptSize = $fieldItems->params['size'];
 			$fileSize = $singleFile['size'];
 
-			if ($fileSize > 15728640)
+			if($acceptSize != "")
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: add pre-commit hooks check indentation

File: site/helpers/tjfields.php [hunk 11]
```diff
@@ -401,19 +404,50 @@ public function uploadFile($singleFile, $insert_obj_file, $file_field_data)
...
+				{
+					$app->enqueueMessage('File size is greater than '. $acceptSize . 'MB' , 'warning');
 
-				return false;
+					return false;
+				}
 			}
+			else
+			{
+				if ($fileSize > 33554432)
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: file size should not be hard coded

File: site/helpers/tjfields.php [hunk 12]
```diff
@@ -401,19 +404,50 @@ public function uploadFile($singleFile, $insert_obj_file, $file_field_data)
...
+				{
+					$app->enqueueMessage('File size is greater than 32MB', 'warning');
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: use language constant

File: site/helpers/tjfields.php [hunk 13]
```diff
@@ -401,19 +404,50 @@ public function uploadFile($singleFile, $insert_obj_file, $file_field_data)
...
-			$validMIMEArray = explode(',', $okMIMETypes);
+					return false;
+				}
+			}
 
+			// Code for file type validation
+			$acceptType = $fieldItems->params['accept'];
+			if (empty($acceptType))
+			{
+				$okMIMETypes    = 'pdf,PDF,doc,DOC,docx,DOCX,xls,XLS,xlsx,XLSX,jpeg,JPEG,png,PNG,jpg,JPG';
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: Why do we need hardcoded file types

File: site/helpers/tjfields.php [hunk 14]
```diff
@@ -424,7 +458,7 @@ public function uploadFile($singleFile, $insert_obj_file, $file_field_data)
 
 			if (!in_array($fileMime, $validMIMEArray))
 			{
-				$app->enqueueMessage('This <strong>' . $title . '</strong> file type is not allowed for file ' . $filename . '.' . $extension, 'warning');
+				$app->enqueueMessage('Invalid file type', 'warning');
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: use language constant

File: site/helpers/tjfields.php [hunk 15]
```diff
@@ -1601,24 +1633,42 @@ public function getMediaUrl($filePath, $extraUrlPrams = '')
...
 			{
 				$extraUrlPrams = '&' . $extraUrlPrams;
 			}
 
-			$b_lk = 'index.php?option=com_tjfields&task=getMedia&fid=';
-			$link = JUri::root() . substr(JRoute::_($b_lk . $fileId . $extraUrlPrams), strlen(JUri::base(true)) + 1);
-
+			// Here, fpht means file encoded path
+			$encodedPath = base64_encode($filePath);
+			$b_lk = 'index.php?option=com_tjfields&task=getMedia&fpht=';
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: use camel case and give a relevant name

**Comment by vaibhavsTekdi** [comment]
Bug #120430 fix: File type acceptance issue, upload & download (uploaded file), new param for file size

Description : Removed the extra lines & spaces, changed the language constant, used the camel cases where it needed etc. in above commit.

will work on below points in next release :

1) Task #123196 : File Field : File size server validation using param or allowed upload max file size & post max size by server

2) Task #123199 File Field : File type server validation & need to set default file type accept criteria

**Pull request merged**